### ISSUE TITLE
Make invisible tanks visible, shooting rotation, zoom fix

### ISSUE THREAD

**Issue by ZfsrGhS953** [terjanq/xDiep.io#8]
Zoom fix - currently the game doesn't draw objects outside of field of view. But server DOES send some information outside of it. So, it would be useful to make the scroll wheel change the REAL zoom, not by scaling objects on the canvas, but by injecting code into d.js. How to do that? Well, if I knew how to inject code into a script, I would try to replace all "/1920.0" and "/1080.0" with a division by a horizontal and vertical zoom variables that are controlled by the userscript. Or changing values of c[377352] and c[377353], which are also related to zoom in some way. As for "shooting rotation", by that I mean changing rotation of Octo Tank by 45 every time it shoots 4 bullets, to shoot 8 bullets in 4 directions, making it a "Quad Twin" instead.

**Comment by hxxr** [comment]
It's definitely possible to change the rotation of a tank. I've found out that the client sends your cursor's location relative to the entire map, and our script can change what the server receives. So all I have to do is make the scale constant, then I can calculate the new coords to be rotated from the tank by 45 degrees, and then send that to the server.

**Comment by ZfsrGhS953** [comment]
That's good, but what about invisibility? Does the server stop updating that tank when it turns completely invisible?

**Comment by terjanq** [comment]
@ZfsrGhS953 Yeah i know that server sends more information to make the game more smooth, it's obvious. 

If you saw my first version of extension then i made all the changes in d.js using some regular expressions. Then dev started detecting my changes by toString for example. If I change something inside d.js it's trivial to shut it down. So I decided to make my extension totally separated from game code. It's easy now to detect and bypass his countermeasure. 

There is nothing special about invisibility. It's just opacity, I can easy change this value that there will be no invisible tanks but it's a cheat so I won't include it.

**Comment by ZfsrGhS953** [comment]
Wait, but zoom is a cheat too then? Why you added it?

**Comment by terjanq** [comment]
I was trying also change the 1080/1920 values but all I achieved is 4:3 aspect which allow you  to see less horizontal and more vertical.

**Comment by terjanq** [comment]
It's not a cheat. I made it because it was easier to me following people (which I didn't see) by mouse, that's all

**Comment by ZfsrGhS953** [comment]
I mean not like 1080/1440, but 2160/3840 (2 times bigger)

**Comment by terjanq** [comment]
Yes I tried it, but it's scaled into maximum values...

**Comment by ZfsrGhS953** [comment]
OK then, had you any luck with changing c[377352] and c[377353] variables in d.js?

**Comment by terjanq** [comment]
Lel there is no constant values inside code, they change every update, so idk what they are.

I tried change them only by manipulating window.innerWidth etc, but then game gets just scaled into 1080/1920, or 4:3

**Comment by ZfsrGhS953** [comment]
I have an idea then. Find "|0)/1920.0" and" |0)/1080.0" in the code and see what's before it. If it's a variable in an array, it's the variable that means the zoom, and then find code that assigns a value to them, and change that code.

**Comment by terjanq** [comment]
The best way to make changes inside d.js would be downloading it inside `head.js` and replace `var c` with `var c; window.foo = c`. Then in head.js setting  interval to check when window.foo is defined and then get a reference to that object and destroy window.foo because script detects if there are any special global variables active. Then you can easy change whatever you want, but it's up to you. Sadly I have no time to do it.

**Comment by terjanq** [comment]
But if you have some problems I can try help

**Comment by terjanq** [comment]
Eh I checked what these variables are, and it is just `window.innerWidth` and `window.innerHeight`.

I Updated head.js (tricky way to get these arrays) and you have your "c array" in `window.__arrays` so you can play with it in console ;)

It is `var c=new global.Int32Array(buffer)` from code.
